doc/error-handling/error_handling_design_spec.md

```diff
@@ -0,0 +1,429 @@
+
+1.1.8 Provide configuration commands to clear the failed objects.
+
+1.1.9 Framework is only responsible for reporting errors. Any retry/rollback operation is the responsibility of applications.
+
+## 1.2 Configuration and Management Requirements
+Provide CLI commands to display and clear ERROR_DB tables.
+
+## 1.3 Supported objects
+Error handling framework supports notifications for the following tables defined in APP_DB:
+
+- ROUTE_TABLE
+- NEIGH_TABLE
+
+The framework can be extended for other tables depending on application requirements. Currently, only ROUTE/NEIGH tables are supported to address the BGP use case.
+
+# 2 Use cases
+
+BGP application relies on RIB failure status to withdraw or advertise the routes. The error handling framework notifies the BGP application about route programming failures. This helps BGP application to withdraw the failed routes that are advertised to the neighbors.
+
+Following diagram describes a high level overview of the BGP use case:
+
+![BGP use case](images/bgp_use_case.png)
+
+# 3 Design
+
+## 3.1 Error Database
+
+As SONIC architecture relies on the use of centralized Redis-database as means of multi-process communication among all subsystems, the framework re-uses the same mechanism to notify errors back to applications.
+
+A new database, ERROR_DB, is introduced to store the details of failed entries/objects corresponding to various tables. The ERROR_DB tables are defined in application friendly format. Applications can register as consumer of ERROR_DB table to receive error notifications, whereas OrchAgent is registered as producer of ERROR_DB table. If the SAI CREATE/SET method fails, Syncd informs OrchAgent using the notification channel of ASIC_DB. OrchAgent is responsible to translate the ASIC_DB notification and store it in ERROR_DB format. It is also responsible to map the SAI specific error codes to SWSS error codes.
+
+For some objects, a notification may be needed even when the object is successfully programmed (SAI API reports SUCCESS). In such scenarios, the framework does not store the object in ERROR_DB (to reduce memory usage), but does notify the registered applications using ERROR_DB notification channel.
+
+Defining a separate error database has the following advantages over extending existing APP_DB tables for error notifications:
+
+- Applications do not register as consumer & producer for same table, which avoids redundant notifications.
+- Flexible, not tied to any other table format.
+- Extensible to all types of errors in the system, not restricted to APP_DB definitions.
+- Efficient, as notifications are limited to failures in the DB.
+- Notification for delete failures can be supported even when corresponding objects are deleted from APP_DB.
+

```

Comment:
I understand the new DB approach, however here is my thought, why can't we have a single DB (APP DB) and separated by namespaces? ex: configured vs applied/error in the same table so that it could be easy to maintain one table.

Reply:
Can you please provide more details here? Are you suggesting that ERROR tables can be stored in APP_DB, and there is no need to create ERROR_DB? We want to avoid modifying the existing ROUTE_TABLE schema in APP_DB - error handling can optionally be disabled and retain the current behavior.